### PR TITLE
Feat(block-execution): add witness execution

### PR DESCRIPTION
**Pull request by mrLSD**

### CODE REVIEW

**Comment by mrLSD** [comment]
@coderabbitai Please perform a full, adversarial audit of this PR. Do not treat this as a normal stylistic code review.

The purpose of this review is to determine whether the implementation is correct, consensus-safe, witness-safe, secure, and efficient for Ethereum block execution in a zkEVM/stateless execution environment.

Please review the entire PR diff and all relevant surrounding code in the repository, not only individual changed lines. Trace data and invariants across modules, especially:

* evm-block-execution
* evm-trie
* evm-trie-bench
* transaction validation/execution
* state transition logic
* system calls
* witness decoding and witness-backed state access
* sparse MPT traversal
* state-root reconstruction/update logic
* block pre/post execution
* hardfork handling
* receipt/request generation
* ancestor / BLOCKHASH handling

The standard for correctness should be Ethereum consensus execution semantics, not merely “the tests pass”.

1. Primary audit objective

Determine whether this PR implements witness-backed block execution correctly enough to be used as the execution path of a zkEVM block executor / stateless Ethereum block verifier.

The implementation must never:

* accept an invalid Ethereum block;
* reject a valid Ethereum block;
* derive an incorrect post-state;
* derive an incorrect state root;
* produce incorrect receipts, logs, requests, gas accounting, or block outputs;
* silently treat an incomplete witness as a proof of absence;
* allow malformed trie data to influence execution as valid state;
* depend on host-only behavior that is inappropriate for a zkVM guest;
* introduce unnecessary asymptotic or constant-factor costs that materially hurt proving performance.

Review this PR under the assumption that all block data and witness data are adversarial inputs.

⸻

2. Ethereum execution / consensus correctness

Audit the implementation against Ethereum execution-layer semantics across all supported forks.

Pay particular attention to interactions between:

* London / EIP-1559
* Merge / PREVRANDAO semantics
* Shanghai / EIP-4895 withdrawals
* Cancun:
    * EIP-4844
    * EIP-4788
* Prague / Electra-related execution changes implemented here:
    * EIP-2935
    * EIP-6110
    * EIP-7002
    * EIP-7251
    * EIP-7691
    * EIP-7825
    * any relevant request-hash / requests behavior

Verify:

* correct fork activation boundaries;
* correct ordering of pre-block, transaction, and post-block operations;
* correct validation-before-execution ordering;
* correct use of header fields;
* correct gas accounting;
* correct blob gas accounting and blob gas price derivation;
* correct withdrawal application;
* correct account-touch semantics;
* correct EIP-161 empty-account deletion semantics;
* correct handling of zero-value withdrawals;
* correct system-call semantics;
* correct receipt construction;
* correct requests construction;
* correct requests ordering;
* correct deposit log parsing;
* correct post-execution header validation;
* correct state-root behavior;
* correct BLOCKHASH behavior.

Do not assume that tests already encode the right behavior. Compare implementation behavior against protocol semantics independently.

⸻

3. Execution ordering

Explicitly reconstruct the expected block execution pipeline and compare it to the implementation.

Check whether execution follows the required ordering approximately along the lines of:

1. ancestor/header validation;
2. sender recovery;
3. block consensus validation;
4. witness/pre-state binding;
5. pre-execution system calls;
6. transaction loop;
7. receipts/log accumulation;
8. post-execution system calls / requests;
9. withdrawals;
10. touched-empty-account cleanup where applicable;
11. final state transition;
12. state-root calculation/update;
13. receipt root / bloom / gas / request validation;
14. post-execution header field validation.

Flag any ordering difference that can produce consensus divergence.

For each system call, verify whether its state changes, logs, return data, gas use, caller semantics, and failure handling match Ethereum execution-spec behavior.

⸻

4. Witness security model

Perform a dedicated witness-soundness audit.

The most important invariant is:

A missing witness node must never become indistinguishable from a valid proof of absence.

Review all account, storage, code, and ancestor reads.

For every possible state access, determine whether the implementation correctly distinguishes:

1. value proven present;
2. value proven absent;
3. proof unavailable because a required node is missing;
4. malformed witness node;
5. malformed account/storage encoding.

Look specifically for cases where APIs such as:

* basic
* storage
* code
* code_hash
* account existence checks
* storage reads

return a default value first and record an error separately.

Verify that every execution path necessarily propagates the recorded witness failure before the block can be accepted.

Look for error-shadowing situations such as:

* missing witness + invalid nonce;
* missing witness + insufficient balance;
* missing witness + EVM revert;
* missing witness + sender validation failure;
* malformed leaf + transaction failure;
* missing code + system call failure.

A malformed/incomplete witness must not accidentally be converted into some unrelated transaction failure.

⸻

5. Witness completeness versus witness minimality

Check that execution requires exactly the data needed to prove the accessed state and no implicit full-state assumptions remain.

Review:

* account proof requirements;
* storage proof requirements;
* code preimages;
* absent-account proofs;
* absent-slot proofs;
* ancestor headers;
* BLOCKHASH;
* system contract state;
* sender account reads;
* destination account reads;
* account creation;
* SELFDESTRUCT behavior if relevant;
* access-list accesses;
* precompile-related state interactions;
* EIP-7702/delegation-related behavior if present in the existing executor.

Check whether a valid minimal witness can execute successfully.

Also check whether irrelevant malformed witness nodes are correctly ignored if they are not on any authenticated path used by execution.

⸻

6. MPT / sparse trie correctness

Perform an adversarial audit of evm-trie/src/sparse*.

Do not only verify ordinary lookup correctness.

Review exact Ethereum Merkle-Patricia Trie canonicality rules, including:

* canonical RLP;
* short versus long RLP forms;
* minimal length encoding;
* trailing bytes;
* truncated input;
* list arity;
* leaf versus extension paths;
* hex-prefix encoding;
* odd/even paths;
* empty paths;
* embedded child rules;
* hashed child rules;
* < 32 byte embedded-node rule;
* root-node exception;
* branch child representation;
* branch terminal values;
* extension-to-extension canonicality;
* unary branch canonicality;
* child hash substitution;
* partial proofs;
* absent-key proofs;
* secure trie paths;
* arbitrary-length key paths used by generic sparse APIs.

Check carefully whether this implementation rejects any structure Ethereum would consider invalid without rejecting valid Ethereum tries.

In particular, independently validate assumptions such as:

“an extension must lead directly to a branch”

and

“a branch must contain at least two occupied entries”

against canonical Ethereum MPT behavior and the exact point at which those invariants are guaranteed.

If these are valid canonicality rules, confirm that enforcement remains correct in partial witnesses where a child subtree is not revealed.

If they are too strict, identify valid proofs that would be rejected.

⸻

7. Trie decoder safety

Audit decode.rs as security-critical untrusted-input parsing code.

Check all indexing and arithmetic for:

* out-of-bounds slices;
* arithmetic overflow;
* truncation;
* usize differences between 32-bit and 64-bit platforms;
* u16 branch offsets;
* large RLP values;
* very large witness nodes;
* recursive embedded nodes;
* pathological nesting;
* denial-of-service behavior.

Confirm that arbitrary witness bytes cannot panic.

Do not rely only on fuzz-like unit tests; reason through all slice/index construction manually.

Pay special attention to expressions derived from:

* bytes.len()
* raw.len()
* payload lengths
* value_start
* path_start
* path_end
* child_start
* branch offset accumulation.

⸻

8. NodeStore construction and sorting

Audit NodeStore::new, sort::by_hash, deduplication, and binary lookup.

Verify:

* sorting preserves each hash ↔ bytes ↔ decoded-metadata association;
* permutation-cycle logic is correct for every permutation;
* duplicate handling cannot retain mismatched bytes or decode metadata;
* hash collisions are handled consistently with the assumed cryptographic model;
* empty/single-element cases are correct;
* sorting does not use invalid sentinels or overflow indices;
* behavior is correct on both 32-bit and 64-bit hosts;
* no architecture-dependent consensus behavior exists.

Look for subtle mutation-order bugs in the in-place permutation implementation.

⸻

9. Account decoding

Audit Ethereum account leaf decoding carefully.

Ethereum state trie accounts are canonically:

[nonce, balance, storageRoot, codeHash]

Check:

* exact list length;
* integer canonicality;
* nonce restrictions;
* balance encoding;
* storage root length;
* code hash length;
* trailing bytes;
* malformed list acceptance;
* interaction with the local TrieAccount type that contains code_version.

Ensure that a five-field internal representation cannot accidentally cause valid witness decoding to become incompatible with Ethereum mainnet account encoding.

⸻

10. Storage decoding

Audit storage trie leaf handling.

Storage values are RLP-encoded integers.

Check:

* zero handling;
* non-zero canonical integer encoding;
* leading-zero rejection;
* over-32-byte values;
* exact conversion to H256;
* absent slot versus malformed slot;
* storage root propagation.

Ensure zero-valued storage never incorrectly materializes into the trie.

⸻

11. Empty-account semantics

This PR changes behavior around empty accounts and explicitly contains logic/tests concerning preserving empty accounts.

Please perform a deep audit here.

Distinguish:

* account exists in pre-state but is EIP-161-empty;
* account is read-only;
* account is touched;
* account receives zero-value transfer;
* account receives zero-value withdrawal;
* account is created;
* account is deleted;
* storage exists under an otherwise empty account;
* account exists in a witness but is not touched;
* account is proven absent.

Verify exact EIP-161 semantics.

Pay special attention to the apparent shift from:

state-root construction filtering empty accounts

toward:

executor removes touched empty accounts and state-root includes every supplied account.

Determine whether this is correct for:

1. complete materialized state;
2. partial/revealed witness state;
3. read-only empty accounts;
4. historically existing empty accounts;
5. impossible/noncanonical Ethereum states.

Flag any possibility of creating a post-state root that differs from Ethereum clients.

⸻

12. Withdrawals

Audit EIP-4895 behavior.

Verify:

* withdrawals only apply from Shanghai onward;
* amounts are Gwei → Wei correctly;
* repeated recipients aggregate correctly;
* arithmetic cannot silently produce a consensus-invalid result;
* zero-value withdrawals have the correct touching semantics;
* touched-empty cleanup happens at the correct stage;
* withdrawal-created accounts behave correctly;
* withdrawals are applied after transactions at the required point.

The use of saturating_add deserves special scrutiny.

Determine whether saturation is consensus-correct or whether overflow should be impossible / handled differently.

⸻

13. EIP-6110 deposit parsing

Audit the parser against the exact execution-spec behavior.

Check:

* deposit contract address configuration;
* mainnet default;
* event topic matching;
* log ordering;
* receipt ordering;
* exact ABI layout;
* offsets;
* lengths;
* padding;
* little-endian fields;
* malformed deposit logs;
* behavior for logs from the correct address/topic with invalid data;
* behavior for unrelated logs;
* request concatenation order;
* correct request byte length.

Independently verify whether requiring one exact canonical ABI representation matches the Ethereum consensus specification.

Also look for DoS risks from many deposit logs or repeated vector growth.

⸻

14. System calls

Audit:

* EIP-4788
* EIP-2935
* EIP-7002
* EIP-7251

Check:

* caller address;
* target address;
* gas limit;
* value;
* fee behavior;
* caller nonce;
* coinbase reward behavior;
* logs;
* revert semantics;
* failure semantics;
* missing-code behavior;
* return-data interpretation;
* correct call timing;
* state rollback behavior.

This comment in particular must be verified against actual behavior:

“A reverted or failed call changes nothing”

Confirm that MemoryStackState, deconstruct(), and subsequent backend.apply(...) actually satisfy that invariant for all ExitReasons.

Check that state from reverted subcalls cannot leak.

⸻

15. Transaction execution

Inspect the refactored executor.rs in full.

Compare its behavior against the implementation before this PR and against Ethereum semantics.

Check all transaction types supported by this repository.

Verify:

* nonce validation;
* EIP-2681;
* intrinsic gas;
* block gas remaining;
* upfront balance checks;
* EIP-1559 fee calculation;
* effective gas price;
* priority fee;
* base-fee burn semantics;
* coinbase credit;
* refund cap;
* blob transaction checks;
* blob gas charging;
* authorization/delegation behavior if applicable;
* sender-code restrictions such as EIP-3607;
* receipts;
* contract creation;
* failed transactions;
* state rollback;
* cumulative gas used.

Look for behavioral differences introduced by the large rewrite even if unrelated to witness support.

⸻

16. BLOCKHASH / ancestors

Audit the witness ancestor-chain design.

Verify:

* trusted pre-state root comes from the correct parent;
* ancestor hashes are cryptographically derived from exact RLP;
* ancestor linkage is validated;
* parent must exist;
* correct block-number continuity;
* correct maximum ancestor window;
* BLOCKHASH returns exactly Ethereum-specified values;
* current/future/out-of-window blocks return zero;
* malformed headers cannot influence execution;
* duplicate or extra headers cannot reorder semantics.

Check interaction between:

* ExecutionWitness.headers
* ancestor validation
* MemoryVicinity.block_hashes
* witness backend block_hash

and ensure there is exactly one authoritative path.

⸻

17. Pre-state root binding

Verify that the witness is actually cryptographically bound to the expected parent state root.

Check all edge cases:

* empty root;
* root node shorter than 32 bytes;
* root missing;
* malformed root;
* unrelated supplied nodes;
* duplicate nodes;
* incomplete paths.

A block must not execute against any state other than the parent header’s committed state.

⸻

18. Post-state computation

This is critical.

Determine exactly how the final post-state root is obtained when the witness only contains a partial state.

Check whether the implementation:

* updates authenticated trie paths correctly;
* retains untouched blinded subtrees;
* handles account insertion;
* handles account deletion;
* handles storage insertion/update/deletion;
* handles storage root changes;
* handles code hash changes;
* handles account creation from absence proofs;
* handles branch compaction;
* handles extension creation/removal;
* handles embedded↔hashed node threshold transitions.

If the implementation materializes only revealed accounts and then computes a normal state root, verify that this cannot accidentally omit untouched state.

If sparse update/root recomputation is implemented elsewhere in the PR, audit that implementation in detail.

If full stateless post-state reconstruction is not actually possible with the current code, identify this as a blocker.

⸻

19. zkEVM / zkVM-specific audit

Review the implementation specifically as code intended to execute inside a zkVM guest.

Do not optimize only for native host execution.

Analyze:

* number of Keccak operations;
* number of RLP decodes;
* repeated hashing;
* binary searches per trie hop;
* witness preprocessing cost;
* sorting cost;
* allocations;
* memory peak;
* data copying;
* temporary buffers;
* branch-heavy logic;
* recursion;
* guest page pressure;
* 32-bit target behavior;
* proving-cycle implications.

The repository mentions RISC Zero / 32-bit guest environments. Evaluate code under that model.

Call out operations that are cheap on a host but expensive in a zkVM.

⸻

20. Performance audit

Give complexity estimates for important paths.

At minimum analyze:

Witness construction/indexing

For:

* N witness nodes
* B total witness bytes

verify the claimed:

* O(B + N log N) construction;
* retained memory;
* temporary sorting memory.

Lookup

For trie depth D and N witness nodes, analyze:

* hash lookup cost;
* binary search cost;
* embedded-node decoding cost;
* allocation count;
* repeated parsing.

Block execution

Estimate cost in terms of:

* transaction count;
* state reads;
* storage reads;
* state writes;
* witness node count;
* receipt/log count;
* deposit count.

Look for adversarial blocks/witnesses that cause disproportionate CPU, memory, or zk proving cost.

⸻

21. DoS analysis

Assume witness and block inputs may be chosen to maximize verifier cost.

Look for:

* oversized witness nodes;
* excessive number of irrelevant nodes;
* excessive duplicate nodes;
* intentionally unsorted witness nodes;
* hash-prefix sorting adversaries;
* deeply nested MPT paths;
* giant branch values;
* many malformed unused nodes;
* repeated misses;
* many log entries;
* huge deposit-request output;
* many contract-code preimages;
* duplicate code;
* pathological transaction access patterns.

Where limits are enforced elsewhere, identify those limits explicitly.

Do not assume upstream limits unless you locate them.

⸻

22. Memory audit

Pay special attention to memory because this is zkVM-sensitive.

Check for:

* duplicate witness storage;
* duplicated RLP bytes;
* duplicate contract code;
* clones of large buffers;
* full materialization of data that could remain borrowed;
* sorting scratch;
* large stack arrays;
* recursive stack usage;
* temporary BTreeMaps;
* Vec over-reservation;
* memory retained after preprocessing.

Distinguish:

* host heap allocations;
* guest pages;
* peak memory;
* retained memory.

⸻

23. 32-bit correctness

The CI explicitly tests 32-bit behavior, and the intended guest may be 32-bit.

Audit every conversion involving:

* usize
* u16
* u32
* u64
* RLP lengths
* offsets
* vector lengths
* trie indexes
* transaction indexes.

Look for behavior that works on x86_64 but overflows, rejects valid data, or panics on a 32-bit guest.

⸻

24. Error handling

Check whether errors preserve the correct distinction between:

* invalid block;
* invalid transaction;
* invalid witness;
* incomplete witness;
* malformed trie;
* malformed account;
* malformed storage;
* failed EVM execution;
* failed protocol system call.

Review error precedence.

For an input with multiple faults, make sure the implementation does not expose a semantic bug because one error masks another.

In particular, missing/malformed witness errors should not be silently swallowed because the EVM substitutes zero/default state.

⸻

25. Test-quality audit

Do not merely count tests.

Evaluate whether tests are independent enough to prove correctness.

Identify cases where implementation and tests share the same assumption or helper and can therefore fail in the same way.

Examples:

* reference trie builder sharing implementation assumptions;
* witness builder using the same encoding types being tested;
* expected values derived using production code;
* tests checking only self-consistency instead of Ethereum-client compatibility.

Suggest missing differential tests.

⸻

26. Differential testing recommendations

Where possible, recommend concrete differential tests against established implementations/spec fixtures, such as:

* Ethereum Execution Specs / EEST;
* geth;
* reth;
* Nethermind;
* Besu;
* Erigon;
* alloy trie implementations;
* independent triehash calculations.

The strongest tests should compare final:

* acceptance/rejection;
* state root;
* receipts root;
* logs bloom;
* gas used;
* requests;
* request hash;
* account/storage values.

For witness tests, suggest generating proofs externally rather than only with the repository’s own test witness builder.

⸻

27. Missing adversarial tests

Please explicitly propose tests for at least:

* missing account branch node;
* missing account leaf;
* missing storage branch;
* missing storage leaf;
* malformed account leaf;
* malformed storage leaf;
* malformed code preimage;
* wrong code hash;
* absent account proof;
* absent storage proof;
* account creation from proven absence;
* storage creation from proven absence;
* touched empty account;
* read-only empty account;
* zero-value transaction;
* zero-value withdrawal;
* repeated withdrawals;
* system-call revert;
* system-call exceptional failure;
* missing system contract code;
* malformed deposit event;
* many deposit events;
* malformed ancestor;
* missing ancestor;
* BLOCKHASH boundary at 256 blocks;
* short MPT root;
* 31/32/33-byte child boundaries;
* extension/branch canonicality;
* embedded child malformed off the selected path;
* hashed malformed child off the selected path;
* duplicate witness nodes;
* witness permutation;
* very large valid RLP values;
* 32-bit length limits.

⸻

28. Compare comments/documentation with behavior

Treat comments as claims that need verification.

For every important claim such as:

* “lookup allocates no heap memory”;
* “reverted system calls change nothing”;
* “unused malformed witness nodes are inert”;
* “every supplied account is included”;
* “executor removes touched empty accounts”;
* “extension must lead to branch”;
* “branch must have at least two occupied entries”;
* “root may be shorter than 32 bytes”;
* “construction performs only one retained allocation”;

verify whether the code actually guarantees it.

Flag misleading or overly strong comments even if code is correct.

⸻

29. API/invariant design review

Look beyond local bugs.

Identify APIs that make consensus bugs easy to introduce later.

Examples:

* APIs returning zero/default value while storing an error out-of-band;
* mutable backend operations whose proof invariants are implicit;
* use of Option where proven absence and unknown state should be distinct types;
* state structures that blur “revealed”, “present”, “absent”, and “dirty”;
* execution APIs that accept an ActiveSpec inconsistent with ChainSpec;
* public constructors that can bypass required validation.

Recommend stronger type-level invariants where useful.

⸻

30. Security severity

Classify each finding as:

* Critical — can accept an invalid block, compute incorrect state, or break witness soundness;
* High — realistic consensus divergence, proof unsoundness, panic/DoS on valid-sized adversarial input, or severe zkVM resource issue;
* Medium — edge-case consensus risk, significant resource issue, incomplete validation, or fragile invariant;
* Low — robustness, maintainability, test gap, misleading invariant/documentation;
* Informational — useful hardening or optimization.

Do not inflate severity.

⸻

31. Required format for every finding

For every concrete issue, provide:

Title

Short description.

Severity

Critical / High / Medium / Low / Informational.

Location

Exact file and function, and lines if possible.

Problem

What is wrong.

Why it matters

Explain specifically in terms of Ethereum execution, witness soundness, consensus, security, or zk proving cost.

Trigger

Give a concrete block/witness/state scenario that reaches the issue.

Expected behavior

What Ethereum/spec-compliant behavior should be.

Current behavior

What this implementation does.

Suggested fix

Prefer a concrete code-level correction.

Regression test

Describe a minimal test that would fail before the fix and pass afterward.

⸻

32. Explicitly separate confirmed issues from questions

Do not report speculative concerns as definite bugs.

Use three sections:

1. Confirmed correctness/security issues
2. Likely issues requiring maintainer confirmation
3. Hardening / performance opportunities

For uncertain protocol interpretations, cite the exact EIP / execution-spec rule that needs clarification.

⸻

33. Final audit summary

At the end, provide a concise summary containing:

* consensus-critical findings;
* witness-soundness findings;
* MPT correctness findings;
* zkVM performance findings;
* missing tests;
* architectural risks.

Also provide a merge-blocker checklist, but do not say merely “LGTM” or “looks safe” based on tests.

The checklist should state objective conditions such as:

* no unresolved consensus divergence;
* incomplete witness can never be interpreted as absence;
* MPT canonicality matches Ethereum;
* post-state root is reproducible from a partial witness;
* all system calls match execution-spec semantics;
* EEST block execution matches;
* 32-bit execution is safe;
* no adversarial witness produces unbounded/patently disproportionate guest memory or compute.

⸻

34. Important review mindset

Assume this code may become part of the trusted execution logic of a zkEVM.

A subtle discrepancy that would normally be considered an edge case is potentially consensus-critical here.

Please prioritize finding:

* semantic bugs over style issues;
* protocol mismatches over refactoring preferences;
* witness unsoundness over ergonomics;
* zkVM costs over native microbenchmark wins.

Ignore formatting/naming/style unless they hide a correctness problem.

Please inspect the entire relevant implementation rather than producing a shallow diff-only review.

File: evm-block-execution/src/stateless.rs [hunk 1]
```diff
@@ -125,124 +143,32 @@ fn stateless_validation_recovered(
...
-            stateless_validation(block, &[], witness, chain_spec()),
-            Err(StatelessValidationError::Consensus(
-                BlockValidationError::DifficultyNotZero { .. }
-            ))
-        ));
-    }
+    Ok(StatelessValidationOutput {
+        block_hash,
+        execution_output,
+    })
```
[the remark this answers is not in the document]
Comment (mrLSD) [reply evm-block-execution/src/stateless.rs]: It's out of scope for this PR

**Comment by mrLSD** [comment]
@coderabbitai provide complex audit. Keep in mind that it's only block wintess execution main goal. Block post validation out of scope.
